### DISCUSSION TITLE
glTF export exclusions

### DISCUSSION THREAD

**Discussion by emperorofmars** [godotengine/godot-proposals#6587] [3D]
Based on Godot 4.0.1

To my knowledge and testing it is not possible to prevent a node from being exported into a glTF node.
For example the 'AnimationPlayer' node, which will become irrelevant once the animations it holds are converted into glTF animations.
That node serves no purpose beyond that, yet it gets exported as a glTF node.
Many custom nodes i have will be saved in a glTF nodes extensions field, but end up being empty nodes in the resulting glTF as well.

I would like to have a way to exclude a Godot node from becoming a glTF node, while still being able to process it on export.
Definetly in 'GLTFDocumentExtension', but perhaps also by a flag on a Godot node itself, maybe in its metadata?
And ideally have that information set on nodes by default where its clear that they should not become glTF nodes, such as 'AnimationPlayer'.

**Comment by lyuma** [comment]
I agree, I ran into this when building the VRM exporter. We have a node called "secondary" with metadata about spring bones that I would like to exclude from export. Generally speaking and node which is not a Node3D probably should be excluded.

Oh, also Godot is exporting the AnimationPlayer node as a node, even though it has no meaning, which breaks round-tripping because the imported document will now name the actual player AnimationPlayer2.

**Reply by lyuma** [reply]
it is possible to rename the node in `_export_post` to something like "unused" (by setting `state.json["nodes"][node_idx]["name"] = ...` ) but it's not a good approach and doesn't prevent leaking nodes in multiple import / export cycles

**Comment by aaronfranke** [comment]
We should probably exclude any non-3D node that does not have children and does not have extension data. This way we will preserve the ability for extensions to keep non-3D nodes exportable, but they will not export by default.
